syntax = "proto3";

package ar_stream;

// Main message envelope
message ARFrame {
  // Metadata
  int64 timestamp_ns = 1;           // Nanosecond timestamp
  uint32 frame_number = 2;          // Sequential frame ID
  string device_id = 8;             // Stable device identifier (e.g., Android ID)

  // Camera data
  CameraData camera = 3;

  // Image data
  ImageFrame rgb_frame = 4;         // Optional: can be omitted for bandwidth

  // Depth data
  DepthFrame depth_frame = 5;

  // Motion/tracking data
  MotionData motion = 6;

  // AR-specific data
  ARCoreData arcore = 7;            // Optional: planes, points, etc.
}

message CameraData {
  // Intrinsic matrix (3x3 flattened to 9 floats)
  repeated float intrinsic_matrix = 1 [packed=true];

  // Projection matrix (4x4 flattened to 16 floats)
  repeated float projection_matrix = 2 [packed=true];

  // View matrix (4x4 flattened to 16 floats)
  repeated float view_matrix = 3 [packed=true];

  // Camera pose (4x4 transform, world space)
  repeated float pose_matrix = 4 [packed=true];

  // Image dimensions
  uint32 image_width = 5;
  uint32 image_height = 6;

  // FOV
  float fov_horizontal_deg = 7;
  float fov_vertical_deg = 8;

  // Camera tracking state
  TrackingState tracking_state = 9;
}

message ImageFrame {
  // Compressed image data (JPEG or raw bytes)
  bytes data = 1;

  // Format specification
  ImageFormat format = 2;
  uint32 width = 3;
  uint32 height = 4;

  // Compression quality (0-100 for JPEG)
  uint32 quality = 5;
}

message DepthFrame {
  // Raw depth data (16-bit values in millimeters)
  bytes data = 1;                   // width * height * 2 bytes

  uint32 width = 2;
  uint32 height = 3;

  // Depth format (RG8 unpacked to uint16)
  DepthFormat format = 4;

  // Depth range
  float min_depth_m = 5;
  float max_depth_m = 6;

  // Confidence map (optional, same dimensions)
  bytes confidence = 7;             // 0-255 per pixel
}

message MotionData {
  // Device pose in world space
  Pose device_pose = 1;

  // Linear velocity (m/s)
  Vector3 linear_velocity = 2;

  // Angular velocity (rad/s)
  Vector3 angular_velocity = 3;

  // Linear acceleration (m/sÂ²)
  Vector3 linear_acceleration = 4;

  // Gravity vector
  Vector3 gravity = 5;

  // Device orientation (quaternion)
  Quaternion orientation = 6;
}

message ARCoreData {
  // Detected planes
  repeated Plane planes = 1;

  // Point cloud (sparse, subsampled)
  PointCloud point_cloud = 2;

  // Light estimation
  LightEstimate light_estimate = 3;

  // Anchors
  repeated Anchor anchors = 4;
}

// Supporting messages

message Pose {
  Vector3 position = 1;
  Quaternion rotation = 2;
}

message Vector3 {
  float x = 1;
  float y = 2;
  float z = 3;
}

message Quaternion {
  float x = 1;
  float y = 2;
  float z = 3;
  float w = 4;
}

message Plane {
  bytes id = 1;                     // UUID
  Pose center_pose = 2;
  float extent_x = 3;
  float extent_z = 4;
  PlaneType type = 5;               // HORIZONTAL_UPWARD, etc.
  repeated Vector3 polygon = 6;     // Boundary vertices
}

message PointCloud {
  repeated float points = 1 [packed=true];  // X,Y,Z,confidence (4 floats per point)
  uint32 point_count = 2;
}

message LightEstimate {
  Vector3 main_light_direction = 1;
  Vector3 main_light_intensity = 2;
  repeated float spherical_harmonics = 3 [packed=true];  // 27 floats
}

message Anchor {
  bytes id = 1;
  Pose pose = 2;
  TrackingState tracking_state = 3;
}

// Enums

enum TrackingState {
  TRACKING_STATE_UNKNOWN = 0;
  NOT_TRACKING = 1;
  LIMITED = 2;
  TRACKING = 3;
}

enum ImageFormat {
  IMAGE_FORMAT_UNKNOWN = 0;
  RGB_888 = 1;
  RGBA_8888 = 2;
  YUV_420 = 3;
  JPEG = 4;
  GRAYSCALE = 5;
}

enum DepthFormat {
  DEPTH_FORMAT_UNKNOWN = 0;
  UINT16_MILLIMETERS = 1;
  FLOAT32_METERS = 2;
}

enum PlaneType {
  PLANE_TYPE_UNKNOWN = 0;
  HORIZONTAL_UPWARD_FACING = 1;
  HORIZONTAL_DOWNWARD_FACING = 2;
  VERTICAL = 3;
}
